api/crossmodelrelations/crossmodelrelations.go

```diff
@@ -28,85 +32,209 @@ func (c *Client) Close() error {
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	if len(results.Results) != 1 {
-		return nil, errors.Errorf("expected 1 result, got %d", len(results.Results))
-	}
-	result := results.Results[0]
-	if result.Error != nil {
-		return nil, result.Error
+	var result params.RelationUnitsWatchResult
+
+	// We make up to 2 api calls - the second is a retry after a macaroon discharge is obtained.
+	for i := 0; i < 2; i++ {
+		var results params.RelationUnitsWatchResults
+		err := c.facade.FacadeCall("WatchRelationUnits", args, &results)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if len(results.Results) != 1 {
+			return nil, errors.Errorf("expected 1 result, got %d", len(results.Results))
+		}
+		result = results.Results[0]
+		if result.Error != nil && i == 0 {
+			mac, err := c.handleError(result.Error)
+			if err != nil {
+				err = errors.Wrap(result.Error, err)
+				return nil, err
+			}
+			args.Args[0].Macaroons = mac
+		}
 	}
 	w := apiwatcher.NewRelationUnitsWatcher(c.facade.RawAPICaller(), result)
 	return w, nil
 }
 
 // RelationUnitSettings returns the relation unit settings for the given relation units in the remote model.
 func (c *Client) RelationUnitSettings(relationUnits []params.RemoteRelationUnit) ([]params.SettingsResult, error) {
-	args := params.RemoteRelationUnits{relationUnits}
-	var results params.SettingsResults
-	err := c.facade.FacadeCall("RelationUnitSettings", args, &results)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	if len(results.Results) != len(relationUnits) {
-		return nil, errors.Errorf("expected %d result(s), got %d", len(relationUnits), len(results.Results))
+	var (
+		args         params.RemoteRelationUnits
+		retryIndices []int
+	)
+
+	result := make([]params.SettingsResult, len(relationUnits))
+	args = params.RemoteRelationUnits{RelationUnits: relationUnits}
+
+	// We make up to 2 api calls - the second is a retry after a macaroon discharge is obtained.
+	for i := 0; i < 2; i++ {
+		var results params.SettingsResults
+		err := c.facade.FacadeCall("RelationUnitSettings", args, &results)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if len(results.Results) != len(args.RelationUnits) {
+			return nil, errors.Errorf("expected %d result(s), got %d", len(args.RelationUnits), len(results.Results))
+		}
+
+		if i == 0 {
+			args = params.RemoteRelationUnits{}
+			for j, res := range results.Results {
+				if res.Error == nil {
+					result[j] = res
+					continue
+				}
+				mac, err := c.handleError(res.Error)
+				if err != nil {
+					res.Error.Message = err.Error()
+					result[j] = res
+					continue
+				}
+				retryArg := relationUnits[j]
+				retryArg.Macaroons = mac
+				args.RelationUnits = append(args.RelationUnits, retryArg)
+				retryIndices = append(retryIndices, i)
+			}
+			if len(args.RelationUnits) == 0 {
+				break
+			}
+		}
+		if i == 1 {
+			for j, res := range results.Results {
+				result[retryIndices[j]] = res
+			}
+		}
 	}
```

Comment:
Given that the handling for iteration 0 and iteration 1 are so different, the code might be clearer if instead of using a for loop, the API calling part was extracted to a separate method and it was called twice. Something like:

* callAPI()
* code to handle the first case, returning early if nothing else to do
* callAPI()
* code to handle second case

This goes for the other API calls too.